File: frontend/lib/repositories/tiploc-repository.ts
import { readFileSync } from 'fs'
import { join } from 'path'

export interface TiplocLocation {
  lat: number
  lng: number
  name: string
}

export class TiplocRepository {
  private readonly tiplocs: Map<string, TiplocLocation>

  constructor() {
    const path = join(process.cwd(), 'data', 'tiplocs.json')
    const raw = JSON.parse(readFileSync(path, 'utf-8')) as Record<string, TiplocLocation>
    this.tiplocs = new Map(Object.entries(raw).map(([k, v]) => [k.toUpperCase(), v]))
  }

  get(tiploc: string): TiplocLocation | null {
    return this.tiplocs.get(tiploc.toUpperCase()) ?? null
  }

  has(tiploc: string): boolean {
    return this.tiplocs.has(tiploc.toUpperCase())
  }
}

File: api/Services/TiplocRepository.cs
using System.Text.Json;

namespace Api.Services;

public record TiplocLocation(double Lat, double Lng, string Name);

public class TiplocRepository
{
    private readonly Dictionary<string, TiplocLocation> _locations;

    public TiplocRepository(IWebHostEnvironment env)
    {
        var path = Path.Combine(env.ContentRootPath, "Data", "tiplocs.json");
        if (File.Exists(path))
        {
            var json = File.ReadAllText(path);
            _locations = JsonSerializer.Deserialize<Dictionary<string, TiplocLocation>>(json,
                             new JsonSerializerOptions { PropertyNameCaseInsensitive = true })
                         ?? new Dictionary<string, TiplocLocation>();
        }
        else
        {
            _locations = new Dictionary<string, TiplocLocation>();
        }
    }

    public TiplocLocation? Get(string tiploc) =>
        _locations.TryGetValue(tiploc, out var loc) ? loc : null;

    public bool Has(string tiploc) => _locations.ContainsKey(tiploc);
}
